metricbeat/module/kafka/consumergroup/query.go

```diff
@@ -38,51 +39,55 @@ func fetchGroupInfo(
 
 	debugf("known consumer groups: ", groups)
 
+	assignments, err := fetchGroupAssignments(b, groups)
+	if err != nil {
+		logp.Err("failed to fetch kafka group assignments: %v", err)
+		return err
+	}
+	if len(assignments) == 0 {
+		return nil
+	}
+
 	wg := sync.WaitGroup{}
-	results := make(chan result, len(groups))
-	for _, group := range groups {
-		group := group
+	results := make(chan result, len(assignments))
+
+	waiting := 0
+	for group, topics := range assignments {
+		_partitions := make(map[string][]int32)
+		for topic, partitions := range topics {
+			if topicsFilter == nil || topicsFilter(topic) {
+				for partitionID := range partitions {
+					if _, ok := _partitions[topic]; !ok {
+						_partitions[topic] = make([]int32, 0)
+					}
+					_partitions[topic] = append(_partitions[topic], partitionID)
+				}
+			}
+		}
+		if len(_partitions) == 0 {
+			continue
+		}
 
 		wg.Add(1)
-		go func() {
+		waiting++
+		go func(group string, partitions map[string][]int32, assign map[string]map[int32]groupAssignment) {
 			defer wg.Done()
-			resp, err := fetchGroupOffset(b, group, topicsFilter)
+			resp, err := fetchGroupOffset(b, group, partitions)
 			if err != nil {
 				logp.Err("failed to fetch '%v' group offset: %v", group, err)
 			}
-			results <- result{err, group, resp}
-		}()
-	}
-
-	go func() {
-		wg.Wait()
-		close(results)
-	}()
-
-	assignments, err := fetchGroupAssignments(b, groups)
-	if err != nil {
-		// wait for workers to stop and drop results
-		for range results {
-		}
-
-		return err
+			results <- result{err, group, assign, resp}
+		}(group, _partitions, topics)
 	}
 
-	for ret := range results {
-		if err := ret.err; err != nil {
-			// wait for workers to stop and drop results
-			for range results {
-			}
-			return err
+	for waiting > 0 {
+		ret := <-results
+		waiting--
+		if ret.err != nil {
+			continue
```

Comment:
originally we did return an error -> we still want to return the first error we encounter
```
for waiting > 0 {
  ret := <- results
  waiting--
  if ret.err != nil {
    if err == nil {
      err = ret.err
    }
    continue
  }

  ...
}
if err != nil {
  return err
}

```

Reply:
fixed: return the error if has